cranelift/codegen/src/machinst/lower.rs

```diff
@@ -942,11 +942,29 @@ impl<'func, I: VCodeInst> Lower<'func, I> {
             // Avoid immutable borrow by explicitly indexing.
             let (inst, succ) = self.vcode.block_order().succ_indices(block)[succ_idx];
             // Get branch args and convert to Regs.
-            let branch_args = self.f.dfg.insts[inst]
-                .branch_destination()
-                .into_iter()
-                .flat_map(|block| block.args_slice(&self.f.dfg.value_lists));
 
+            // TODO:
+            // let branch_args = self.f.dfg.insts[inst]
+            //     .branch_destination()
+            //     .into_iter()
+            //     .flat_map(|block| block.args_slice(&self.f.dfg.value_lists));
+
+            let branch_args = match self.f.dfg.analyze_branch(inst) {
+                instructions::BranchInfo::NotABranch => unreachable!(),
+                instructions::BranchInfo::SingleDest(block) => {
+                    block.args_slice(&self.f.dfg.value_lists)
+                }
+                instructions::BranchInfo::Conditional(then_block, else_block) => {
+                    // TODO: does succ_idx actually index the block of the instruction?
```

Comment:
I think it does so in the order returned by `visit_branch_targets` which you implemented as `then_block` and then `else_block`, but the code to compute this doesn't seem to be made to preserve this order, but rather preserved it by accident. See https://github.com/bytecodealliance/wasmtime/blob/94b51cdb179d08dbedd2934a3dc11e2c5aef481a/cranelift/codegen/src/machinst/blockorder.rs